### ISSUE TITLE
Deploy script fails for packages that depend on other packages

### ISSUE THREAD

**Issue by kujtimprenkuSQA** [bikecoders/ngx-deploy-npm#541]
In an NX monorepo with multiple packages, the deploy script is failing for packages that depend on other packages.

GitHub workflow errors: https://github.com/near/wallet-selector/actions/runs/5952823742/job/16211406235#step:12:51

Steps to reproduce:

- Clone https://github.com/near/wallet-selector.git
- yarn && yarn build:all
- yarn nx run-many --target=deploy --exclude=react,angular --dry-run --verbose

ngx-deploy-npm version: 7.0.1

NX version: 16.0.0
Node version: 18.17.1
Yarn version: 1.22.19
OS: any

**Comment by dianjuar** [comment]
I don't believe it's a deployment problem but a build problem.

If you execute the `build` target `yarn nx run-many --target=build --exclude=react,angular  --verbose` do you get the same error?

**Comment by kujtimprenkuSQA** [comment]
> I don't believe it's a deployment problem but a build problem.
> 
> If you execute the `build` target `yarn nx run-many --target=build --exclude=react,angular --verbose` do you get the same error?

No, there are no errors when we execute the build, you might see in the above workflow we have a "Run Build All" where we attempt to build all packages before running the "deploy" script.

Would the result be the same if we build all the packages with the `build` target and then run the `deploy` script with the `--noBuild` flag?

**Comment by dianjuar** [comment]
> Would the result/deployment be the same if we build all the packages with the build target and then run the deploy script with the --noBuild flag?

I think that might work.

It is an ngx-deploy-npm issue that is prepared to build their dependent packages.

---

I'll suggest using [`--postTargets`](https://github.com/jscutlery/semver#triggering-executors-post-release) to publish to NPM if there is something to publish. If there is no new version, the deployment to NPM will fail because you are attempting to publish an already-published version.

**Comment by kujtimprenkuSQA** [comment]
Thank you for your reply.

We found out why the build was failing when executing the `deploy` script without the `--noBuild` flag.

For some reason even if we execute the `build` target before the `deploy` the `ngx-deploy-npm:deploy` does not look into the `dist/packages/package-name` to check if it exists but tries to do a fresh build on the package, at this point, the build is done in `parallel` so the packages which are set as dependencies might not have been built yet.

Adding `deploy` to `targetDefaults` in `nx.json` makes sure that dependencies are built and deployed first:

```json
  "targetDefaults": {
    "deploy": {
      "dependsOn": ["^deploy"]
    }
  },
```

In our use-case we decided to handle the build separately and added the `--noBuild` flag when we execute the `deploy` target.

**Comment by dianjuar** [comment]
That makes sense.

Another workaround would be adding `dependsOn` on the deploy target.

```json
{
  "targets": {
    "deploy": {
      "executor": "ngx-deploy-npm:deploy",
      "options": {
        "access": "public"
      },
      "dependsOn": ["^build"]
    }
  }
}
```

**Comment by kujtimprenkuSQA** [comment]
Closing this issue it has been resolved in our side. Thank you for the help.

**Comment by dianjuar** [comment]
> Adding `deploy` to `targetDefaults` in `nx.json` makes sure that dependencies are built and deployed first:
> 
> ```json
>   "targetDefaults": {
>     "deploy": {
>       "dependsOn": ["^deploy"]
>     }
>   },
> ```
> 
> In our use-case we decided to handle the build separately and added the `--noBuild` flag when we execute the `deploy` target.

Do you leave this on your project?

Since your library depends on another library, the dependencies must be published to NPM first. I understand that Nx doesn't add other libraries in your bundle. I was experimenting the other day with [incremental builds](https://nx.dev/concepts/more-concepts/incremental-builds) and found out that Nx doesn't add other libraries in your bundle, you are forced to publish the dependencies to NPM.

This issue comment will add more light to my comment -> https://github.com/nrwl/nx/issues/4620#issuecomment-1156858443

**Comment by kujtimprenkuSQA** [comment]
Yes, we left the `deploy` target in `targetDefaults` in `nx.json` for now it's merged only in our [dev](https://github.com/near/wallet-selector/tree/dev) branch, I marked this issue as "resolved" mostly because now we're not seeing build errors happening, we have not tried deploying the packages yet as we still have the `--dry-run` flag on.

Historically we have manually published our packages by executing the [release-packages.bash](https://github.com/near/wallet-selector/blob/main/scripts/release-packages.bash) script.

Now after reading your comment and the comments on the links you shared, I am not sure if we should just keep publishing the packages manually or attempt to automate it with NX.
We will have to investigate this a bit more to find out if everything is working correctly.
